src/dmd/dcast.d

```diff
@@ -3827,3 +3835,58 @@ IntRange getIntRange(Expression e)
     e.accept(v);
     return v.range;
 }
+
+Type checkArrayOpType(Expression e1, Expression e2, TOK op, Scope *sc)
+{
+    Expression lhs = e1;
+    Expression rhs = e2;
+
+    // scalar op scalar - we shouldn't be here
+    if (lhs.type.ty != Tarray && lhs.type.ty != Tsarray && rhs.type.ty != Tarray && rhs.type.ty != Tsarray)
+        return null;
+
+    if (!lhs.isSliceExp() && !lhs.isArrayLiteralExp() && !rhs.isSliceExp() && !rhs.isArrayLiteralExp())
+        return null;
+
+    // T[x .. y] op ?
+    if (e1.isSliceExp())
```

Comment:
it's confusing to use both e1 and lhs. Choose one and stick with it.

Reply:
At the end of the execution I want to restore e1 and e2. At least one of them will be changed inside the function. I rearranged the code a bit, I hope now it makes more sense.